src/d3a/models/strategy/update_frequency.py

```diff
@@ -15,247 +15,141 @@
-        self._decrease_price_every_nr_s = 0
-        self.initial_selling_rate = initial_selling_rate
-
-    def assign_offermixin_arguments(self, initial_rate_option, energy_rate_decrease_option,
-                                    energy_rate_decrease_per_update):
-        if initial_rate_option is not None:
-            self.initial_rate_option = InitialRateOptions(initial_rate_option)
-        if energy_rate_decrease_option is not None:
-            self.energy_rate_decrease_option = RateDecreaseOption(energy_rate_decrease_option)
-        if energy_rate_decrease_per_update is not None:
-            if energy_rate_decrease_per_update < 0:
-                raise ValueError("Energy rate decrease per update should be a positive value.")
-            self.energy_rate_decrease_per_update = energy_rate_decrease_per_update
-
-    def update_on_activate(self):
-        # This update of _decrease_price_every_nr_s can only be done after activation as
-        # MAX_OFFER_TRAVERSAL_LENGTH is not known at construction
-        self._decrease_price_every_nr_s = \
-            (self.area.config.tick_length.seconds *
-             ConstSettings.GeneralSettings.MAX_OFFER_TRAVERSAL_LENGTH + 1)
-
-    def calculate_initial_sell_rate(self, current_time_h):
-        if self.initial_rate_option is InitialRateOptions.HISTORICAL_AVG_RATE:
-            if self.area.historical_avg_rate == 0:
-                return self.area.config.market_maker_rate[current_time_h]
+            self.final_rate[market.time_slot] = final_rate
+        if fit_to_limit is not None:
+            self.fit_to_limit = fit_to_limit
+        if energy_rate_change_per_update is not None:
+            self.energy_rate_change_per_update[market.time_slot] = \
+                energy_rate_change_per_update
+        if update_interval is not None:
+            self.update_interval = update_interval
+
+        self.update_on_activate(strategy)
+
+    def _get_energy_rate_change_per_update(self):
+        energy_rate_change_per_update = {}
+        for slot in generate_market_slot_list():
+            if self.fit_to_limit:
+                energy_rate_change_per_update[slot] = \
+                    (self.initial_rate[slot] - self.final_rate[slot]) / \
+                    self.number_of_available_updates
             else:
-                return self.area.historical_avg_rate
-        elif self.initial_rate_option is InitialRateOptions.MARKET_MAKER_RATE:
-            return self.area.config.market_maker_rate[current_time_h]
-        elif self.initial_rate_option is InitialRateOptions.CUSTOM_RATE:
-            return self.initial_selling_rate
+                energy_rate_change_per_update[slot] = self.energy_rate_change_per_update[slot]
+        return energy_rate_change_per_update
+
+    def _calculate_number_of_available_updates_per_slot(self, strategy):
+        number_of_available_updates = \
+            int(strategy.area.config.slot_length.seconds / self.update_interval.seconds) - 1
```

Comment:
Did you check whether with this calculation the final_rate is really reached (even for combinations of e.g. slot_length=60s and update_interval=7s)?

Reply:
Sorry for the broken link, have a look at this link (https://github.com/gridsingularity/d3a/blob/0f81e5fdd70d5502498c9c2dc8d7147d71a04049/src/d3a/models/strategy/update_frequency.py#L78). If I recall correctly, @spyrostz  suggested to have an additional assertion to avoid too low `update_interval` in order to avoid spamming the blockchain with very high frequency of offers/bid update.

We can also discuss offline too